File: backend/src/models/Visit.js
const mongoose = require('mongoose');

// Une visite = entree dans le geofence d'un client jusqu'a sa sortie
const visitSchema = new mongoose.Schema(
  {
    agent: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    client: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Client',
      required: true,
      index: true,
    },
    enteredAt: { type: Date, required: true, index: true },
    leftAt: { type: Date },
    durationSeconds: { type: Number },
    entryLocation: {
      type: { type: String, enum: ['Point'], default: 'Point' },
      coordinates: [Number],
    },
    exitLocation: {
      type: { type: String, enum: ['Point'], default: 'Point' },
      coordinates: [Number],
    },
    status: {
      type: String,
      enum: ['in_progress', 'completed', 'too_short', 'too_long', 'cancelled'],
      default: 'in_progress',
    },
    // ===== Gestion du chevauchement de geofences =====
    // Quand plusieurs clients se chevauchent, la visite demarre automatiquement
    // sur le client le plus proche mais reste "non confirmee" tant que l'agent
    // n'a pas valide (ou corrige) le client reellement visite.
    isConfirmed: { type: Boolean, default: true, index: true },
    candidateClients: [
      { type: mongoose.Schema.Types.ObjectId, ref: 'Client' },
    ],
    notes: String,
  },
  { timestamps: true }
);

visitSchema.methods.endVisit = function (exitLocation) {
  this.leftAt = new Date();
  this.durationSeconds = Math.round((this.leftAt - this.enteredAt) / 1000);
  if (exitLocation) this.exitLocation = exitLocation;
  this.status = 'completed';
  return this.save();
};

module.exports = mongoose.model('Visit', visitSchema);

File: mobile/lib/models/visit.dart
/// Modele visite client
class Visit {
  final String id;
  final String agentId;
  final String clientId;
  final String? clientName;
  final String? clientType;
  final DateTime enteredAt;
  final DateTime? leftAt;
  final int? durationSeconds;
  final String status; // in_progress, completed, too_short, too_long, cancelled
  final String? notes;

  Visit({
    required this.id,
    required this.agentId,
    required this.clientId,
    this.clientName,
    this.clientType,
    required this.enteredAt,
    this.leftAt,
    this.durationSeconds,
    required this.status,
    this.notes,
  });

  factory Visit.fromJson(Map<String, dynamic> json) {
    String agentId = '';
    String clientId = '';
    String? clientName;
    String? clientType;

    if (json['agent'] is String) {
      agentId = json['agent'];
    } else if (json['agent'] is Map) {
      agentId = json['agent']['_id'] ?? '';
    }

    if (json['client'] is String) {
      clientId = json['client'];
    } else if (json['client'] is Map) {
      clientId = json['client']['_id'] ?? '';
      clientName = json['client']['name'];
      clientType = json['client']['type'];
    }

    return Visit(
      id: json['_id'] ?? '',
      agentId: agentId,
      clientId: clientId,
      clientName: clientName,
      clientType: clientType,
      enteredAt: DateTime.parse(json['enteredAt']),
      leftAt: json['leftAt'] != null ? DateTime.parse(json['leftAt']) : null,
      durationSeconds: json['durationSeconds'],
      status: json['status'] ?? 'in_progress',
      notes: json['notes'],
    );
  }

  bool get isOngoing => status == 'in_progress';

  String get formattedDuration {
    if (durationSeconds == null) return '-';
    final s = durationSeconds!;
    if (s < 60) return '${s}s';
    final m = s ~/ 60;
    final r = s % 60;
    return '${m}m ${r}s';
  }
}
